shell/common/animator.cc

```diff
@@ -65,7 +70,17 @@ void Animator::BeginFrame(ftl::TimePoint frame_time) {
   // TODO(abarth): We should use |frame_time| instead, but the frame time we get
   // on Android appears to be unstable.
   last_begin_frame_time_ = ftl::TimePoint::Now();
+  // We use less than 1/60 second to account for delays between the vsync
+  // notification and this calculation.
+  dart_frame_deadline_ = Dart_TimelineGetMicros() + 15000;
+
   engine_->BeginFrame(last_begin_frame_time_);
+
+  if (!frame_scheduled_) {
+    // We don't have another frame pending, so we're waiting on user input
+    // or I/O. Allow the Dart VM 100 ms.
+    engine_->NotifyIdle(dart_frame_deadline_ + 100000);
```

Comment:
What happens when the VM runs out of time?

Reply:
The VM does not start a GC if it estimates it cannot complete it before the deadline. If the estimate is wrong, the VM will exceed the deadline (GC is not interruptible). If the VM overruns a deadline, this would still provide a better frame rate because the next frame is only delayed by the amount of the overrun instead of the full duration of a GC in the middle of a frame.

Reply:
Cool, ok. We should document that somewhere. Maybe on `Window.onBeginFrame` for now? You could add a `## Interaction with garbage collection` subsection to that dartdoc block, say.